rate_limit.go

```diff
@@ -5,49 +5,38 @@ import (
 	"time"
 )
 
+// Inspired by http://stackoverflow.com/a/668327
 type rateLimiter struct {
-	v int64
+	rate, limit int64
 
-	_closed int64
+	allowance, lastCheck int64
 }
 
-func newRateLimiter(limit time.Duration, bucketSize int) *rateLimiter {
-	rl := &rateLimiter{
-		v: int64(bucketSize),
+func newRateLimiter(limit time.Duration, rate int) *rateLimiter {
+	if limit < time.Second {
+		limit = time.Second
 	}
-	go rl.loop(limit, int64(bucketSize))
-	return rl
-}
+	return &rateLimiter{
+		rate:  int64(rate) * 1e9,
+		limit: int64(limit),
 
-func (rl *rateLimiter) loop(limit time.Duration, bucketSize int64) {
-	for {
-		if rl.closed() {
-			break
-		}
-		if v := atomic.LoadInt64(&rl.v); v < bucketSize {
-			atomic.AddInt64(&rl.v, 1)
-		}
-		time.Sleep(limit)
+		allowance: int64(rate) * 1e9,
+		lastCheck: time.Now().UnixNano(),
 	}
 }
 
-func (rl *rateLimiter) Check() bool {
-	for {
-		if v := atomic.LoadInt64(&rl.v); v > 0 {
-			if atomic.CompareAndSwapInt64(&rl.v, v, v-1) {
-				return true
-			}
-		} else {
-			return false
-		}
-	}
-}
+func (rl *rateLimiter) Limit() bool {
```

Comment:
Please document return value since it is not obvious what true/false means in this context.